runner/runner.go

```diff
@@ -0,0 +1,282 @@
+
+const unitStatusChannelSize = 10
+
+const (
+	defaultStartTimout         = 30
+	unitStartTimeoutMultiplier = 2
+)
+
+const systemdUnitNameTamplate = "aos-service@%s.service"
+
+const errNotLoaded = "not loaded"
+
+const (
+	failedState   = "failed"
+	activeState   = "active"
+	jobeStateDone = "done"
+)
+
+/***********************************************************************************************************************
+ * Vars
+ **********************************************************************************************************************/
+
+var ErrStartInstance = errors.New("can't start service instance")
+
+/***********************************************************************************************************************
+  Types
+ **********************************************************************************************************************/
+
+// StartInstanceParams start instance parameters.
+type StartInstanceParams struct {
+	StartInterval   time.Duration
+	StartBurst      uint
+	RestartInterval time.Duration
+}
+
+// InstanceStatus service instance status.
+type InstanceStatus struct {
+	InstanceID string
+	State      string
+	Err        error
+	ExitCode   int
+}
+
+// Runner runner instance.
+type Runner struct {
+	systemd            *dbus.Conn
+	subSet             *dbus.SubscriptionSet
+	instanceStatusChan chan []InstanceStatus
+	startingUnits      map[string]chan dbus.UnitStatus
+	stopChan           chan bool
+}
+
+/***********************************************************************************************************************
+  Public
+ **********************************************************************************************************************/
+
+// New creates new systemd runner.
+func New() (runner *Runner, err error) {
+	runner = &Runner{
+		instanceStatusChan: make(chan []InstanceStatus, unitStatusChannelSize),
+		startingUnits:      make(map[string]chan dbus.UnitStatus),
+		stopChan:           make(chan bool, 1),
+	}
+
+	// Create systemd connection
+	runner.systemd, err = dbus.NewSystemConnectionContext(context.Background())
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	runner.subSet = runner.systemd.NewSubscriptionSet()
+
+	go runner.monitorUnitSates()
+
+	return runner, nil
+}
+
+// Close closes runner.
+func (runner *Runner) Close() {
+	log.Debug("Close runner")
+
+	runner.stopChan <- true
+
+	runner.systemd.Close()
+}
+
+func (runner *Runner) InstanceStatusChannel() <-chan []InstanceStatus {
+	return runner.instanceStatusChan
+}
+
+// StartInstance start service instance as systemd unit.
+func (runner *Runner) StartInstance(instanceID, runtimeDir string, params StartInstanceParams) (status InstanceStatus) {
+	status.InstanceID = instanceID
+	status.State = failedState
+
+	unitName := fmt.Sprintf(systemdUnitNameTamplate, instanceID)
+
+	runner.startingUnits[unitName] = make(chan dbus.UnitStatus)
+
+	runner.subSet.Add(unitName)
```

Comment:
It is not thread-safe to add/remove items to subSet after subscription. I'm afraid you should write your own subscription implementation based on SubscribeUnitsCustom.

Reply:
reworked